### ISSUE TITLE
The state of extensions

### ISSUE THREAD

**Issue by JuBan1** [notepadqq/notepadqq#686]
So, it's clear that not much happens with extensions. I think, none of us even know for certain whether they still work or if we've borked them at some point along the road. I know I couldn't get them to work.

The main issue for me as a Nqq developer is the difficulty of working on it. We've got external processes, [APIs](https://github.com/notepadqq/NotepadqqApi_Nodejs) to bridge them to Nqq, and internal C++ code that isn't quite obvious either. Honestly, that makes me very hestitant to even attempt to improve and/or fix it.

So I've been thinking about a possible replacement that does away with all of it: instead of using NodeJS to host an extension's JS environment we could use Qt's own JavaScript engine to do so. That means we could:

* Get rid of NodeJS and NPM as dependencies, as well as the accompanying APIs.
* Tighten up our code significantly.
* Use Qt signals as a means of communication between Nqq and extensions.
* Actually get something done.

But it comes with some significant drawbacks, namely:

* As far as I can gather, Qt doesn't use a "browser.js"-like environment, unlike NodeJS. That means many common JS libraries that depend on this environment can't simply be imported into our extensions. A number of common features (like timers) would probably have to be re-created by us.

* We will need to depend on QtQml as a new build-time dependency. Although I think this is available basically anywhere. We might need a minimum of Qt 5.5 or 5.6 to get all the necessary functionality. Not sure about this one, though.

* Without NPM we would need our own simple packaging/management solution for extension packages. Although this is probably not a big issue.

That first bullet point is the most troubling. We'd lose a lot of potential. For example: if someone were to create an extension that synchronizes open files in Nqq with an FTP server, that would probably be somewhat simple with our current approach using the proper JS libraries, but a huge pain with the approach I suggested since those libraries wouldn't work.

I'm not happy with that trade-off but I think it's better than the status quo.

Regarding the actual implementation: I've been looking at two (and a half) alternatives

* QtScript would fit the bill nicely but is deprecated. :frowning_face: 
* QJSEngine is not as powerful as QtScript but works. Needs the qml module as a dependency.
* QQmlEngine is what's used for running Qml code in non-qtwidget applications. Qml is a markup language for creating GUIs but it has the ability to run fully-fledged JS code. It's somewhat of a misappropriation but I found it to work rather well. It integrates very well into Nqq using signals and brings along its own set of tools (like timers) that we won't need to reinvent. I wrote a tiny proof-of-concept program to illustrate how it would work that I can upload.

**Long story short:** What will we do with extensions? Sticking to the current system seems like leaving them to rot. Going with my suggestion means lowering our aspirations in exchange for actual progress. I'm a complete amateur with everything JavaScript- perhaps there are other alternatives I don't know.

Thoughts?

**Comment by danieleds** [comment]
So, I think a few things are important.

 * As you already said, a "browser.js"-like environment for max compatibility with existing libraries.
 * Independence from notepadqq when doing network requests or file system access or shell access (e.g. nodejs could do all of them without bothering notepadqq ("async") )
 * Abstraction. Our internal code changes continuously: I believe in the need for a stable abstraction layer (like we had, with an API, or in any other way).

Node and NPM aren't bad *per se*... in the current state yes, but maybe we can find a better, seamless integration.
I'm not familiar with QQmlEngine or QJSEngine but if they work that's good. Can you show your proof of concept?

**Comment by danieleds** [comment]
I remember that initially I tried using python, but it was impossible to have proper self-contained packages so at the end js was chosen. The fact that the editor itself is js is a bonus that allows further integration.

**Comment by JuBan1** [comment]
@danieleds 

I'll post the POC when I get home and polished it up a little bit. To address your points:

* browser.js is, like I said, my sore spot. I don't think Qt's JS Engine is in any way compatible. I've been scouring the internet for a solution without success. But then again, my knowledge about the JS infrastructure is limited.

* Independence shouldn't be hard. I think JS runs on its own thread by default, and if we add our own objects they won't have to interact with Nqq code unless we want it to.

* Abstraction is no problem. But it would be a few C++ objects that act as the interface between QJSEngine and Nqq.

@Teklad 

It doesn't *have* to be JS but that's the only scripting language Qt supports out of the box. Any other language would mean additional dependencies. I wouldn't couple extensions tightly to our WebEngine instance though. While it would probably work it would give us an awful time if we ever want to transition away from WebEngine.

**Comment by JuBan1** [comment]
Ok, here's the proof of concept: https://github.com/JuBan1/JSTest

It's a tiny Qt program. After building it you'll have to manually copy the 'test.js' and 'package.json' files into the build directory (so they're in the same dir as the executable). The code only looks for an extension (if you can call it that..) in that directory.

It's a whitespace killer. Type into the textbox and press the button to have the extension remove all whitespace from the text.

**Comment by JuBan1** [comment]
It's ES5, with some work being done to bring support up to ES7. It supports the usual way of doing web requests:

    var http = new XMLHttpRequest()
    http.open("POST", url);
    http.setRequestHeader(...)
    http.onreadystatechange = function() { /* do stuff */ }

That can also be used to read local (text) files. Anything more fancy and we would probably have to offer our own solution. (The upside to this is that it's pretty easy- you can basically create a class, add some `Q_INVOKABLE`s and slap it into the QJSEngine for everyone to use.)

**Comment by scx** [comment]
I've just added extension support to my flatpak package.
https://github.com/scx/notepadqq-flatpak/commit/ba9e0ecc0a17ae6d8f5dc7655d2b51b43926018d

As you can see, I bundled modified version of the `node` _0.12.x_, because the project state that it is the latest supported version.
https://github.com/notepadqq/notepadqq/blob/9f8ee0baeefa60f6af15ddfeb969614c38115999/src/ui/frmpreferences.ui#L710
Anyway, it is completely independent for the host. I can bundle whatever version you want. Just tell me which version works best for you. It could be _0.10.x_, _0.11.x_, _0.12.x_, ..., _8.x_, _10.x_, _12.x_, etc.

After some minor [difficulties](https://github.com/qcerasmus/notepadqq-extensions/issues/3), I was able to install `Hex2Ascii`.
`Beautifier` almost works for me, but I had to [remove](https://github.com/qcerasmus/notepadqq-extensions/issues/7) some code.
I don't know if it is because of incompatible version of Notepadqq or Node.js.

I managed to created two [sample extensions](https://github.com/scx/notepadqq-extensions), so generally speaking, it works.

![notepadqq-emoticons](https://user-images.githubusercontent.com/510369/65886101-046a7600-e39c-11e9-80ac-a1180503b765.png)

![notepadqq-currencies](https://user-images.githubusercontent.com/510369/65886119-0b918400-e39c-11e9-82f8-5b05bdbde922.png)

I just need to know what version of node you want to officially support.
_0.10.x_, _0.11.x_, _0.12.x_ are really outdated. They don't even support ECMAScript 2015 (ES6).
https://node.green/
However, if this is your choice, I'm fine with it.

**Comment by scx** [comment]
See also:
https://github.com/qcerasmus/notepadqq-extensions/issues/8
https://github.com/qcerasmus/notepadqq-extensions/issues/9

**Comment by danieleds** [comment]
It's good to know that extensions still work :)
With all the major changes that happened over the last year I was expecting them to be completely broken.

I'd say that we should upgrade to supporting the latest LTS version of node. I believe that old extensions *may* work out of the box with the latest nodejs, but we'd have to check and anyway the instructions should change according to current best practices

**Comment by scx** [comment]
> It's good to know that extensions still work :)
> With all the major changes that happened over the last year I was expecting them to be completely broken.
> 
> I'd say that we should upgrade to supporting the latest LTS version of node. I believe that old extensions _may_ work out of the box with the latest nodejs, but we'd have to check and anyway the instructions should change according to current best practices

The problem with the new version of nodejs is that it will require some changes in the existing code. We can't just provide e.g. node _8.x_, _10.x_ or _12.x_, because it won't work.

For example, currently Notepadqq try to pass the `--harmony-proxies` option to the node, but it was removed in node _7.x_.
https://github.com/notepadqq/notepadqq/blob/v2.0.0-beta/src/ui/Extensions/extension.cpp#L38

And even with node _6.x_, it won't work, because `Proxy.create` is not a function with node _6.x_.
```
TypeError: Proxy.create is not a function
```

Someday we will eventually have to remove `proxy` support and add `promise` support, but it will take some time.
For this reason, I think it is a task for [v2.1](https://github.com/notepadqq/notepadqq/milestone/4).

See also:
https://github.com/ovh/node-ovh/pull/10/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1e
